### ISSUE TITLE
LeakSanitizer has encountered a fatal error

### ISSUE THREAD

**Issue by fgsch** [google/sanitizers#1306]
clang-10 -v:

`Ubuntu clang version 10.0.1-++20200708122807+ef32c611aa2-1~exp1~20200707223407.61`

Repro:

```
$ echo "int main(int argc, char **argv) { }" >repro.c
$ clang-10 -fsanitize=address repro.c -o repro
$ sudo chgrp root repro
$ sudo chmod g+s repro
$ ./repro
```

The group in step 3 (chgrp) can be any group the user does not belong to.

Error:

```
==5662==LeakSanitizer has encountered a fatal error.
==5662==HINT: For debugging, try setting environment variable LSAN_OPTIONS=verbosity=1:log_threads=1
==5662==HINT: LeakSanitizer does not work under ptrace (strace, gdb, etc)
```
`ASAN_OPTIONS=detect_leaks=0` is ignored. 
Adding `LSAN_OPTIONS=verbosity=1:log_threads=1` doesn't add any details.

**Comment by fgsch** [comment]
Also, this can be reproduced with older clang versions, e.g., `clang version 5.0.1 (tags/RELEASE_501/final)`.

**Comment by kdurnoga** [comment]
You might try setting `LSAN_OPTIONS=use_tls=0`. If that works then the issue is likely to be the same as indicated in #1342.

**Comment by fgsch** [comment]
Sorry, but no, it does not work.

The problem here is that the Linux kernel (at least) will not allow you to ptrace a process if the effective group id does not match the real group id (same for user). 

I suppose the real problem is that there is no way to disable LSAN in such cases.

**Comment by kdurnoga** [comment]
Yep, you're right. Although you can disable LSAN even in this case using sanitizer hooks. Just define
```
int __lsan_is_turned_off() {
  return 1;
}
```
This should be provided inside your binary though, LD_PRELOAD won't probably work due to its limitations on setuid/setgid binaries.

**Comment by eugenis** [comment]
ASAN_OPTIONS=detect_leaks=0 is ignored because /proc/self/environ can not
be opened in a setguid binary, EACCES.

**Comment by GavinRay97** [comment]
@kdurnoga Bless you, putting this in my C++ `main` fixed it for me:

```cmake
target_compile_definitions(${PROJECT_NAME}
        PRIVATE
        $<$<CONFIG:DEBUG>:DEBUG>)
```

```cpp
#ifdef DEBUG
extern "C" int __lsan_is_turned_off()
{
    return 1;
}
#endif
```

**Comment by matthiasgreuter** [comment]
Any updates here? I also encounter this problem.
